From b9ecd90b0bda46bffeaf2bc515cdd335f6c570c5 Mon Sep 17 00:00:00 2001
From: Huang Shijie <b32955@freescale.com>
Date: Thu, 23 Dec 2010 10:55:56 +0800
Subject: [PATCH] ENGR00137270-1 GPMI : add a separate DMA starting function with BCH interrupt

Add a separate DMA starting function for common use.

Imx23 and imx28 should not wait for the DMA without time out.
If a bug exits in the DMA engine, it will very hard to debug.

Signed-off-by: Huang Shijie <b32955@freescale.com>
---
 drivers/mtd/nand/gpmi-nfc/gpmi-nfc-hal-common.c |   32 +++++++++++++++++++++++
 drivers/mtd/nand/gpmi-nfc/gpmi-nfc.h            |    1 +
 2 files changed, 33 insertions(+), 0 deletions(-)

diff --git a/drivers/mtd/nand/gpmi-nfc/gpmi-nfc-hal-common.c b/drivers/mtd/nand/gpmi-nfc/gpmi-nfc-hal-common.c
index e5ba51f..7aacc89 100644
--- a/drivers/mtd/nand/gpmi-nfc/gpmi-nfc-hal-common.c
+++ b/drivers/mtd/nand/gpmi-nfc/gpmi-nfc-hal-common.c
@@ -428,6 +428,38 @@ int gpmi_nfc_dma_go(struct gpmi_nfc_data *this, int  dma_channel)
 
 }
 
+/*
+ * This function is used in BCH reading or BCH writing.
+ * It will wait for the BCH interrupt as long as one second.
+ * Actually, we will wait for two interrupts, the DMA interrupt and
+ * BCH interrupt.
+ *
+ * @this:        Per-device data structure.
+ * @dma_channel: DMA channel
+ *
+ */
+int start_dma_with_bch_irq(struct gpmi_nfc_data *this, int dma_channel)
+{
+	struct nfc_hal *nfc = this->nfc;
+	int error = 0;
+
+	/* Prepare to receive an interrupt from the BCH block. */
+	init_completion(&nfc->bch_done);
+
+	/* Go! */
+	error = gpmi_nfc_dma_go(this, dma_channel);
+	if (error)
+		printk(KERN_ERR "[ %s ] DMA error\n", __func__);
+
+	/* Wait for the interrupt from the BCH block. */
+	error = wait_for_completion_timeout(&nfc->bch_done,
+						msecs_to_jiffies(1000));
+	error = (!error) ? -ETIMEDOUT : 0;
+	if (error)
+		printk(KERN_ERR "[ %s ] bch timeout!!!\n", __func__);
+	return error;
+}
+
 /**
  * ns_to_cycles - Converts time in nanoseconds to cycles.
  *
diff --git a/drivers/mtd/nand/gpmi-nfc/gpmi-nfc.h b/drivers/mtd/nand/gpmi-nfc/gpmi-nfc.h
index dfaccd1..8cd8ab2 100644
--- a/drivers/mtd/nand/gpmi-nfc/gpmi-nfc.h
+++ b/drivers/mtd/nand/gpmi-nfc/gpmi-nfc.h
@@ -609,6 +609,7 @@ extern int gpmi_nfc_set_geometry(struct gpmi_nfc_data *this);
 extern int gpmi_nfc_dma_go(struct gpmi_nfc_data *this, int  dma_channel);
 extern int gpmi_nfc_compute_hardware_timing(struct gpmi_nfc_data *this,
 					struct gpmi_nfc_hardware_timing *hw);
+extern int start_dma_with_bch_irq(struct gpmi_nfc_data *this, int dma_channel);
 
 /* NFC HAL Structures */
 
-- 
1.5.4.4

